Restructure bot contributing guide (#591)
discord-python/site: pydis_site/apps/content/resources/guides/pydis-guides/contributing/bot.md

See [here](../obtaining-discord-ids) for help with obtaining Discord ID's.

<div class="card">
    <button type="button" class="card-header collapsible">
        <span class="card-header-title subtitle is-6 my-2 ml-2">Optional config.yml</span>
        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">
              <p>If you used the provided server template, and you're not sure which channels belong where in the config file, you can use the config below. Pay attention to the comments with several <code>#</code> symbols, and replace the <code>�</code> characters with the right ID's.</p>
              <pre>
                <code class="language-yaml">
bot:
    prefix:      "!"

    redis:
        host:  "redis"
        password: null
        port:  6379
        use_fakeredis: true

    stats:
        presence_update_timeout: 300
        statsd_host: "graphite.default.svc.cluster.local"

urls:
    # PyDis site vars
    site:        &DOMAIN       "web:8000"

>     site_api:    &API    !JOIN ["api.", *DOMAIN]

Reviewer: ```suggestion
    site_api:    &API    !JOIN [*DOMAIN, "/api"]
```
This change is needed since we removed subdomains

    site_api_schema:           "http://"
    site_paste:  &PASTE  !JOIN ["paste.", "pythondiscord.com"]
    site_schema: &SCHEMA       "http://"

>     site_staff:  &STAFF  !JOIN ["staff.", *DOMAIN]

Reviewer: ```suggestion
    site_staff:  &STAFF  !JOIN [*DOMAIN, "/staff"]
```
This change is needed since we removed subdomains

Note that if you changed code that is not associated with a particular cog, such as utilities, converters, and constants, you will need to restart the bot.

#### Adding new statistics

Details on how to add new statistics can be found on the [statistic infrastructure page](https://blog.pythondiscord.com/statistics-infrastructure).
We are always open to more statistics so add as many as you can!

---

### Optional: Working with Redis
In [Configure the Bot](#configyml) you were asked to set `bot.redis.use_fakeredis` to `true`. If you do not need to work on features that rely on Redis, this is enough. Fakeredis will give the illusion that features relying on redis are saving information properly, but restarting the bot or the specific cog will wipe that information.

If you are working on a feature that relies on redis, you will need to enable redis to make sure persistency is achieved for the feature across restarts. The first step towards that is going to `config.yml` and setting `bot.redis.use_fakeredis` to `false`.

#### Starting Redis in Docker (Recommended)

> If you're using the Docker image provided in the project's docker-compose, open your `config.yml` file, set `bot.redis.host` to `redis`, and `bot.redis.password` to `null`.

Reviewer: ```suggestion
If you're using the Docker image provided in the project's docker-compose, open your `config.yml` file, set `bot.redis.host` to `redis`, `bot.redis.password` to `null`, and `bot.redis.use_fakeredis` to `false`.
```

Author: This is stated right above the subtitle.

title: Contributing to Bot
description: A guide to setting up and configuring Bot.
icon: fab fa-github
toc: 3
---
The purpose of this guide is to get you a running local version of [the Python bot](https://github.com/python-discord/bot).
This page will focus on the quickest steps one can take, with mentions of alternatives afterwards.

### Clone The Repository
First things first, to run the bot's code and make changes to it, you need a local version of it (on your computer).

<div class="card">
    <button type="button" class="card-header collapsible">
        <span class="card-header-title subtitle is-6 my-2 ml-2">Getting started with git and GitHub</span>
        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">

>               <p>If you don't have git on your computer already, <a href="https://git-scm.com/downloads">install it</a>. You can additionally install a git GUI such as <a href="https://www.gitkraken.com/download">GitKraken</a>, or the <a href="https://cli.github.com/manual/installation">GitHub CLI</a>.</p>

Reviewer 2: All references to "Git" should be capitalised. It's currently inconsistent.

              <p>To learn more about git, you can look into <a href="../working-with-git">our guides</a>, as well as <a href="https://education.github.com/git-cheat-sheet-education.pdf">this cheatsheet</a>, <a href="https://learngitbranching.js.org">Learn Git Branching</a>, and otherwise any guide you can find on the internet. Once you got the basic idea though, the best way to learn git is to use it.</p>

>               <p>Creating a copy of a repository under your own account is called a <em>fork</em>. This is where all your changes and commits will be pushed to, and from where your pull requests will originate from. For any staff member, since you have write permissions already to the original repository, you can just create a feature branch to push your commits to instead.</p>

Reviewer 2: The last sentence is redundant since the same info is stated in the "Working on the Repository Directly" section.

#### Working on the Repository Directly
If you are a member of the organisation (a member of [this list](https://github.com/orgs/python-discord/people), or in our particular case, server staff), you can clone the project repository without creating a fork, and work on a feature branch instead.

---

### Set Up a Test Server
The Python bot is tightly coupled with the Python Discord server, so to have a functional version of the bot you need a server with channels it can use.
It's possible to set the bot to use a single channel for all cogs, but that will cause extreme spam and will be difficult to work with.

You can start your own server and set up channels as you see fit, but for your convenience we have a template for a development server you can use: [https://discord.new/zmHtscpYN9E3](https://discord.new/zmHtscpYN9E3).
Keep in mind that this is not a mirror of the Python server, but a reduced version for testing purposes. A lot of the channels in the Python server were merged.

---

### Set Up a Bot Account
You will need your own bot account on Discord to test your changes to the bot.

> See [here](../creating-bot-account) for help with setting up a bot account, and invite your bot to the test server you created in the previous section.

Reviewer 2: There is something wrong with this sentence. I think you meant to write "..., and invit**ing** your bot ..."

Author: Hmm the second part of the sentence is not related to the guide, it's what you're supposed to do once you have a bot account. I'll try to phrase it better.

The Python bot requires the `Server Member Intent` to function.
In order to enable it, visit the [Developer Portal](https://discord.com/developers/applications/) (from where you copied your bot's login token) and scroll down to the `Privileged Gateway Intents` section.
The `Presence Intent` is not necessary and can be left disabled.

If your bot fails to start with a `PrivilegedIntentsRequired` exception, this indicates that the required intent was not enabled.

---

### Configure the Bot
You now have both the bot's code and a server to run it on. It's time you to connect the two by changing the bot's configurations.

#### config.yml
Entering the directory of the cloned code, you will find a file named `config-default.yml`.
This file contains the various configurations we use to make the bot run on the Python Discord server, such as channel and role ID's, and the emojis it works with.
It also contains configurations such as how long it takes for a help channel to time out, and how many messages a user needs to voice-verify.

To run the bot in your test server, you will need to override some of those configurations.

> Open a new file in the directory called `config.yml`. Alternatively, copy the `config-default.yml` file and rename the copy to `config.yml`.

Reviewer 2: To be clearer, this should state "create and open a new file".

* Set `urls.site_api` to `!JOIN [*DOMAIN, "/api]`.
* Set `urls.site_schema` and `urls.site_api_schema` to `"http://"`.

We understand this is tedious and are working on a better solution for setting up test servers.

<div class="card">
    <button type="button" class="card-header collapsible">
        <span class="card-header-title subtitle is-6 my-2 ml-2">Why do you need a separate config file?</span>
        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">
            While it's technically possible to edit <code>config-default.yml</code> to match your server, it is heavily discouraged.
            This file's purpose is to provide the configurations the Python bot needs to run in the Python server in production, and should remain as such.
            In contrast, the <code>config.yml</code> file can remain in your local copy of the code, and will be ignored by commits via the project's <code>.gitignore</code>.
        </div>
    </div>
</div>
<br>

#### .env

> The second file you need to open is the one containing the environment variables, and needs to have the name `.env`.

Reviewer 2: Clarify that the file should be created if it doesn't exist (which is most likely the case for anyone reading the guide). Also, I think "need to be named" reads a bit better than "needs to have the name".

Now that you have everything setup, it is finally time to make changes to the bot!

#### Working with Git

If you have not yet [read the contributing guidelines](../contributing-guidelines), now is a good time.
Contributions that do not adhere to the guidelines may be rejected.

Notably, version control of our projects is done using Git and Github.
It can be intimidating at first, so feel free to ask for any help in the server.

[**Click here to see the basic Git workflow when contributing to one of our projects.**](../working-with-git/)

#### Running tests

[This section](https://github.com/python-discord/bot/blob/main/tests/README.md#tools) of the README in the `tests` repository will explain how to run tests.
The whole document explains how unittesting works, and how it fits in the context of our project.

Make sure to run tests *before* pushing code.

#### Lint before you push
As mentioned in the [contributing guidelines](../contributing-guidelines), you should make sure your code passes linting for each commit you make.

For ease of development, you can install the pre-commit hook with `poetry run task precommit`, which will check your code every time you try to commit it.

> For that purpose, even if you run the bot through docker, you might want to [setup a development environment](#setting-up-a-development-environment), as otherwise the hook installation will fail.

Reviewer 2: Should be capitalised anywhere else it's currently lowercase too (if there are any other instances).

```suggestion
For that purpose, even if you run the bot through Docker, you might want to [setup a development environment](#setting-up-a-development-environment), as otherwise the hook installation will fail.
```

#### Reloading parts of the bot

> If you make changes to a cog, you might not need to restart the entire bot for the changes to take effect. The command `!cog reload <cog_name>` (or `!ext r <cog_name>`, or `!c r <cog_name>` in short) re-imports the files associated with the cog.

Reviewer 2: Using "cog" as terminology here should really be discouraged. It's reloading the whole extension, not just the cog class. Furthermore, the command takes the extension's name, not the cog's name. IMO the "cog" aliases shouldn't even be mentioned at all. They only even exist for "backwards compatibility".

Note that if you changed code that is not associated with a particular cog, such as utilities, converters, and constants, you will need to restart the bot.

#### Adding new statistics

Details on how to add new statistics can be found on the [statistic infrastructure page](https://blog.pythondiscord.com/statistics-infrastructure).
We are always open to more statistics so add as many as you can!

---

### Optional: Working with Redis

> In [Configure the Bot](#configyml) you were asked to set `bot.redis.use_fakeredis` to `true`. If you do not need to work on features that rely on Redis, this is enough. Fakeredis will give the illusion that features relying on redis are saving information properly, but restarting the bot or the specific cog will wipe that information.

Reviewer 2: Make sure Redis is capitalised everywhere.
```suggestion
In [Configure the Bot](#configyml) you were asked to set `bot.redis.use_fakeredis` to `true`. If you do not need to work on features that rely on Redis, this is enough. Fakeredis will give the illusion that features relying on Redis are saving information properly, but restarting the bot or the specific cog will wipe that information.
```

> If you are working on a feature that relies on redis, you will need to enable redis to make sure persistency is achieved for the feature across restarts. The first step towards that is going to `config.yml` and setting `bot.redis.use_fakeredis` to `false`.

Reviewer 2: ```suggestion
If you are working on a feature that relies on Redis, you will need to enable Redis to make sure persistency is achieved for the feature across restarts. The first step towards that is going to `config.yml` and setting `bot.redis.use_fakeredis` to `false`.
```

#### Starting Redis in Docker (Recommended)
If you're using the Docker image provided in the project's docker-compose, open your `config.yml` file, set `bot.redis.host` to `redis`, and `bot.redis.password` to `null`.

#### Starting Redis Using Other Methods
You can run your own instance of Redis, but in that case you will need to correctly set `bot.redis.host` and `bot.redis.port`, and the `bot.redis.password` value in `config-default.yml` should not be overridden. Then, enter the `.env` file, and set `REDIS_PASSWORD` to whatever password you set.

---

### Optional: Working with Metricity

> [Metricity](https://github.com/python-discord/metricity) is our home-grown bot for collecting metrics on activity within the server, such as what users are  present, and ID's of the messages they've sent.

Reviewer 2: The plural of "ID" is "IDs". What you're currently writing is the possessive form rather than the plural. There are several instances of this mistake that need to be corrected.

Certain features in the Python bot rely on querying the Metricity database for information such as the number of messages a user has sent, most notably the voice verification system.

If you wish to work on a feature that relies on Metricity, for your convenience we've made the process of using it relatively painless with docker: Enter the `.env` file you've written for the Python bot, and append the line `USE_METRICITY=true`.
Note that if you don't need Metricity, there's no reason to have it enabled as it is just unnecessary overhead.

> To make the metricity bot work with your test server, you will need to override its configurations similarly to the Python bot.

Reviewer 2: Check capitalisation anywhere else this word is used.
```suggestion
To make the Metricity bot work with your test server, you will need to override its configurations similarly to the Python bot.
```

You can see the various configurations in [the Metricity repo](https://github.com/python-discord/metricity), but the bare minimum is the guild ID setting.
In your local version of the Python bot repo, create a file called `metricity-config.toml` and insert the following lines:
```yaml
[bot]
guild_id = replace_with_your_guild_id
```

> To properly replicate production behavior, set the `staff_role_id`, `staff_categories` and `ignore_categories` fields as well.

Reviewer 2: Serial comma — this is a hill I will die on
```suggestion
To properly replicate production behavior, set the `staff_role_id`, `staff_categories`, and `ignore_categories` fields as well.
```

title: Contributing to Bot
description: A guide to setting up and configuring Bot.
icon: fab fa-github
toc: 3
---
The purpose of this guide is to get you a running local version of [the Python bot](https://github.com/python-discord/bot).
This page will focus on the quickest steps one can take, with mentions of alternatives afterwards.

### Clone The Repository
First things first, to run the bot's code and make changes to it, you need a local version of it (on your computer).

> Clone the repository from [here](https://github.com/python-discord/bot). If you are not a member of the organisation, you will need to create a fork of the project, and clone the fork instead.

Reviewer 3: I think we can specify what a `member of the organisation` is, since that may not be clear for people new to Github.
Minor comment though, feel free to take it or leave it.

Author: Reworded that section, is that better?

### Set Up a Bot Account
You will need your own bot account on Discord to test your changes to the bot.
See [here](../creating-bot-account) for help with setting up a bot account, and invite your bot to the test server you created in the previous section.

#### Privileged Intents

It is necessary to explicitly request that your Discord bot receives certain gateway events.
The Python bot requires the `Server Member Intent` to function.
In order to enable it, visit the [Developer Portal](https://discord.com/developers/applications/) (from where you copied your bot's login token) and scroll down to the `Privileged Gateway Intents` section.
The `Presence Intent` is not necessary and can be left disabled.

If your bot fails to start with a `PrivilegedIntentsRequired` exception, this indicates that the required intent was not enabled.

### Configure the Bot

> You now have both the bot's code, and a server to run it on. Now you need to connect the two by changing the bot's configurations.

Reviewer 3: Two `nows` in quick succession. Sounded odd to me.
```suggestion
You now have both the bot's code and a server to run it on. It's time to connect the two by changing the bot's configurations.
```

#### config.yml
Entering the directory of the cloned code, you will find a file named `config-default.yml`.
This file contains the various configurations we use to make the bot run on the Python Discord server, such as channel and role ID's, and the emojis it works with.
It also contains configurations such as how long it takes for a help channel to time out, and how many messages a user needs to voice-verify.

To run the bot in your test server, you will need to override some of those configurations.
Open a new file in the directory called `config.yml`. The bot will first look at the items in this file, and will fall back to `config-default.yml` only if necessary.

For example, `config-default.yml` contains the item:
```yaml
bot:
    prefix:         "!"
```
If you add the following lines to `config.yml`:
```yaml
bot:
    prefix:         "~"
```

> The bot will use `~` for its prefix instead.

Reviewer 3: I'm not sure how necessary this example is. I think it's enough to say that "if something is present on the config.yml, it will override what is in the default-config.yml for that entry". 

This example just separates the two options of:
1. Create new file, copy and paste the provided template
2. Duplicate default-config.yml and change values

so upon reading it seems like there's only 1 option.

Author: Shifted things around, is that clearer?

    # before message making a channel dormant
    deleted_idle_minutes: 1

    # Maximum number of channels to put in the available category
    max_available: 2

    # Maximum number of channels across all 3 categories
    # Note Discord has a hard limit of 50 channels per category, so this shouldn't be > 50
    max_total_channels: 20

    # Prefix for help channel names
    name_prefix: 'help-'

    # Notify if more available channels are needed but there are no more dormant ones
    notify: true

    # Channel in which to send notifications
    notify_channel: *DEV_LOUNGE_HELPERS_ORG

    # Minimum interval between helper notifications
    notify_minutes: 5

    # Mention these roles in notifications
    notify_roles:
        - *HELPERS_ROLE

##### <<  Add any additional sections you need to override from config-default.yml  >> #####
            </code>
          </pre>
</div></div></div>
<br>

If you don't wish to use the provided `config.yml` above, you can copy the `config-default.yml` file, rename it, and change whatever values you need. Note that you don't need to specify all items in `config.yml`, just the ones you want to override such as channel ID's. 

> See [here](../obtaining-discord-ids) for help with obtaining Discord ID's.

Reviewer 3: Can we make this standout with some sort of callout? It's currently getting lost in between all the text and I think it would be helpful to mark this as a "this is useful info! Check it out if you need it".

Something like this: 
![image](https://user-images.githubusercontent.com/20641196/134241302-b6c846eb-e015-4cfe-b131-4ac9f484a5f8.png)
(ref: https://pythondiscord.com/pages/guides/pydis-guides/contributing/sir-lancebot/)

Author: I highlighted it, but it's still inside the collapsible. You think it should be outside?

These are the main sections in `config-default.yml` that need overriding:

* `guild.id`
* `guild.categories`
* `guild.channels`
* `guild.roles`
* `guild.webhooks`
* `style.emojis`

Additionally:

* Set `urls.site_schema` and `urls.site_api_schema` to `"http://"`.
* Set `urls.site_api` to `!JOIN ["api.", *DOMAIN]`.
* Set `urls.snekbox_eval_api` to `"http://localhost:8060/eval"`.

> * At this stage, set `bot.redis.use_fakeredis` to `true`. If you're looking for instructions for working with Redis, see [Working with Redis](#optional-working-with-redis).

Reviewer 3: Can we list these in the order that they appear in the config.yml? Small, but helpful change.

Also cc: @ChrisLovering I remember you told me to do extra stuff here once upon a time during the nightmare of setting up Python. I think it's related to this section?

> We understand this is tedious and are working on a better solution for setting up test servers.

Reviewer 3: I don't think we need this line. We've had it for awhile now and no progress to the bootstrapping process has been made. I'd opt to remove it until we actually have a better way of getting this setup.

Author: We *are* working on a solution though :P

Reviewer 3: We've been working on a solution for like... a year though >_>

Author: Still, I'd rather people didn't think we're not aware of this.

        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">
            While it's technically possible to edit <code>config-default.yml</code> to match your server, it is heavily discouraged.
            This file's purpose is to provide the configurations the Python bot needs to run in the Python server in production, and should remain as such.
            In contrast, the <code>config.yml</code> file can remain in your local copy of the code, and will be ignored by commits via the project's <code>.gitignore</code>.
        </div>
    </div>
</div>
<br>

#### .env
The second file you need to open is the one containing the environment variables, and needs to have the name `.env`. 
Inside, add the line `BOT_TOKEN=YourDiscordBotTokenHere`. See [here](../creating-bot-account) for help with obtaining the bot token.

The `.env` file will be ignored by commits.

### Run it!
#### With Docker
You are now almost ready to run the Python bot. The simplest way to do so is with Docker. 

> In your `config.yml` file, set `urls.site` to `"web:8000"`.

Reviewer 3: So, I think we should be default assume people are using Docker and move this up to the previous section talking about config.yml. It's a bit odd to think we're done with config.yml only to have it randomly pop up here again.

I think when looking at the running the bot locally, you can specify that the config.yml has to be different than specified above.

Author: My issue with this is that I still want the document to be a quick reference guide. If someone wants to see how to start it specifically with Docker after trying another method, it should tell them the exact configurations necessary for it.

Reviewer 3: Yeah that's fair

<div class="card">
    <button type="button" class="card-header collapsible">
        <span class="card-header-title subtitle is-6 my-2 ml-2">Getting started with Docker</span>
        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">
            The requirements for Docker are:
            <ul>
                <li><a href="https://docs.docker.com/install">Docker CE</a></li>
                <li>Docker Compose. If you're using macOS and Windows, this already comes bundled with the previous installation. Otherwise, you can download it either from the <a href="https://docs.docker.com/compose/install">website</a>, or by running <code>pip install docker-compose</code>.</li>
            </ul>
            <p>If you get any Docker related errors, reference the <a href="../docker#possible-issues">Possible Issue</a> section of the Docker page.</p>
        </div>
    </div>
</div>
<br>

> Assuming you have Docker installed, enter the cloned repo in the command line and type `docker-compose up`.

Reviewer 3: Can we also specify that you need to have Docker running? It's a small thing that seems obvious, but I think it'd be helpful to specify.

* In your `config.yml` file, set `urls.site` to `"pythondiscord.local:8000"`. If you wish to keep using `web:8000`, then [COMPOSE_PROJECT_NAME](../docker/#compose-project-names) has to be set to use this domain.

You will need to start the services separately, but if you got the previous section with Docker working, that's pretty simple:

* `docker-compose up web` to start the site container. This is required.
* `docker-compose up snekbox` to start the snekbox container. You only need this if you're planning on working on the snekbox cog.
* `docker-compose up redis` to start the Redis container. You only need this if you're not using fakeredis. For more info refer to [Working with Redis](#optional-working-with-redis).

You can start several services together: `docker-compose up web snekbox redis`.

##### Setting Up a Development Environment
The bot's code is Python code like any other. To run it locally, you will need the right version of Python with the necessary packages installed:

1. Make sure you have [Python 3.9](https://www.python.org/downloads/) installed. It helps if it is your system's default Python version.

> 2. Install Poetry by following [these instructions](https://github.com/python-poetry/poetry#installation), or with `pip install poetry`. In case of the latter, make sure pip installs to the right Python version.

Reviewer 3: I recommend not including the `pip install poetry` since it's something poetry itself discourages.

3. [Install the dependencies](../installing-project-dependencies).

> Assuming you have at least the site running in Docker already, you can now start the bot locally through the command line, or through your preferred IDE. Notice that the bot is started as a module.

Reviewer 3: This is a bold assumption. I feel like we should at least link to the site setup instructions at this point so they can go down that rabbit hole if they want.

Also "Notice that the bot is started as a module". I *think* I know what you mean, but I wouldn't expect someone to by default know what that means. I think having the command to start the bot here (above the collapsible section) would be helpful and then the collapsible section can go into further details. Or move the comment into the collapsible section.

Author: This section is directly under the part which mentions how to start the site container (line 521). Do you think it should be mentioned again?

Reviewer 3: Ah, I missed it in the above paragraph. Maybe change it to: "The site has to have been started beforehand, see the above paragraph for the site contributing guide."

Also, for that above paragraph I would maybe change it to:
"You can run additional services on the host. This guide won't go over how to install and start the other dependencies on the host. If possible, prefer to start the services through Docker to replicate the production environment as much as possible.

The site, however, is a mandatory service for the bot. Refer to the [site contributing guide](linky link) to learn how to start it on the host, in which case you will need to change `urls.site` in `config.yml` to wherever the site is being hosted."

Author: I almost committed "linky link" <_<

                <li>If using PyCharm, enter <code>Edit Configurations</code> and set everything according to this image: <img src="/static/images/content/contributing/pycharm_run_module.png"></li>
                <li>If using Visual Studio Code, set the interpreter to the poetry environment you created. In <code>launch.json</code> create a new Python configuration, and set the name of the program to be run to <code>bot</code>. VSC will correctly run it as a module.</li>
            </ul>
        </div>
    </div>
</div>
<br>

#### With More Things Running Locally
You can run additional services on the host. For the site, refer to the [site contributing guide](../site) to learn how to start it on the host, in which case you will need to change `urls.site` in `config.yml` to wherever the site is being hosted. This guide won't go over how to install and start the other dependencies on the host. If possible, prefer to start the services through Docker to replicate the production environment as much as possible.

#### Start Only the Bot with Docker

> This method will start only the bot using Docker. The site has to have been started somehow beforehand.

Reviewer 3: Like having the site run in a separate docker instance? What other options are there?
This feels odd enough to not include without further explanation.

Start the bot using Docker Compose while inside the root of the project directory: `docker-compose up --no-deps bot`.

### Development Tips
Now that you have everything setup, it is finally time to make changes to the bot!

#### Working with Git

If you have not yet [read the contributing guidelines](../contributing-guidelines), now is a good time.
Contributions that do not adhere to the guidelines may be rejected.

Notably, version control of our projects is done using Git and Github.
It can be intimidating at first, so feel free to ask for any help in the server.

[**Click here to see the basic Git workflow when contributing to one of our projects.**](../working-with-git/)

#### Running tests

[This section](https://github.com/python-discord/bot/blob/main/tests/README.md#tools) of the README in the `tests` repository will explain how to run tests.
The whole document explains how unittesting works, and how it fits in the context of our project.

#### Reloading parts of the bot

> If you make changes to a cog, you might not need to restart the entire bot for the changes to take effect. The command `!cog reload <cog_name>` (or `!c r <cog_name>` in short) re-imports the files associated with the cog.

Reviewer 3: `!cog reload <cog_name>` is not correct, it's `!cogs reload <cog_name>`
... Either way I prefer `ext`: `!ext r <name>` so can we include that too? :D

### Appendix: Full ENV File Options
The following is a list of all available environment variables used by the bot:

| Variable | Required | Description |
| -------- | -------- | -------- |
| `BOT_TOKEN` | Always | Your Discord bot account's token (see [Set Up a Bot Account](##set-up-a-bot-account)). |
| `BOT_API_KEY` | When running bot without Docker | Used to authenticate with the site's API. When using Docker to run the bot, this is automatically set. By default, the site will always have the API key shown in the example below. |
| `BOT_SENTRY_DSN` | When connecting the bot to sentry | The DSN of the sentry monitor. |
| `BOT_TRACE_LOGGERS ` | When you wish to see specific or all trace logs | Comma separated list that specifies which loggers emit trace logs through the listed names. If the ! prefix is used, all of the loggers except the listed ones are set to the trace level. If * is used, the root logger is set to the trace level. |
| `BOT_DEBUG` | In production | `true` or `false`, depending on whether to enable debug mode, affecting the behavior of certain features. `true` by default.

> | `REDIS_PASSWORD` | When not using FakeRedis | The password to connect to the redis database. *Leave empty if you're not using REDIS.* |

Reviewer 3: what happens if you don't but still set `fakeredis=true`?

Author: It will have no effect. I guess it's not clear enough

<div class="card">
    <button type="button" class="card-header collapsible">
        <span class="card-header-title subtitle is-6 my-2 ml-2">Getting started with Docker</span>
        <span class="card-header-icon">
            <i class="fas fa-angle-down title is-5" aria-hidden="true"></i>
        </span>
    </button>
    <div class="collapsible-content">
        <div class="card-content">
            The requirements for Docker are:
            <ul>
                <li><a href="https://docs.docker.com/install">Docker CE</a></li>
                <li>Docker Compose. If you're using macOS and Windows, this already comes bundled with the previous installation. Otherwise, you can download it either from the <a href="https://docs.docker.com/compose/install">website</a>, or by running <code>pip install docker-compose</code>.</li>
            </ul>
            <p class="notification is-warning">If you get any Docker related errors, reference the <a href="../docker#possible-issues">Possible Issue</a> section of the Docker page.</p>
        </div>
    </div>
</div>
<br>

In your `config.yml` file:

> * Set `urls.site` to `"web:8000"`.

Reviewer 4: I was setting up an instance of the bot earlier today and I've found setting urls.site to "web:8000" broke it:

```
bot_1        | Traceback (most recent call last):
bot_1        |   File "/bot/bot/api.py", line 66, in maybe_raise_for_status
bot_1        |     response_json = await response.json()
bot_1        |   File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1097, in json
bot_1        |     raise ContentTypeError(
bot_1        | aiohttp.client_exceptions.ContentTypeError: 0, message='Attempt to decode JSON with unexpected mimetype: text/html', url=URL('http://api.web:8000/healthcheck')
bot_1        | 
bot_1        | During handling of the above exception, another exception occurred:
bot_1        | 
bot_1        | Traceback (most recent call last):
bot_1        |   File "/bot/bot/bot.py", line 251, in login
bot_1        |     await self.ping_services()
bot_1        |   File "/bot/bot/bot.py", line 99, in ping_services
bot_1        |     await self.api_client.get("healthcheck")
bot_1        |   File "/bot/bot/api.py", line 80, in get
bot_1        |     return await self.request("GET", endpoint, raise_for_status=raise_for_status, **kwargs)
bot_1        |   File "/bot/bot/api.py", line 75, in request
bot_1        |     await self.maybe_raise_for_status(resp, raise_for_status)
bot_1        |   File "/bot/bot/api.py", line 70, in maybe_raise_for_status
bot_1        |     raise ResponseCodeError(response=response, response_text=response_text)
bot_1        | bot.api.ResponseCodeError: Status: 404 Response: <!DOCTYPE html>
```

```suggestion
* Set `urls.site` to `"web:8000/api"`.
```

@HassanAbouelela suggested "web:8000/api" which seems to work. Not sure if this is exclusive to me, but I thought I'd mention it.

Author: The subdomains were removed from the site a few days ago, so I need to update the guide 👍
